blockscout-ens/bens-logic/src/protocols/protocoler.rs

```diff
@@ -283,16 +292,100 @@ impl Protocoler {
         Ok(protocols)
     }
 
+    pub fn fetch_domain_options(
+        &self,
+        name_with_tld: &str,
+        network_id: i64,
+        maybe_filter: Option<NonEmpty<String>>,
+    ) -> Result<Vec<DomainNameOnProtocol>, ProtocolError> {
+        let tld = self.extract_tld(name_with_tld)?;
+
+        let protocols = self.protocols_of_network_for_tld(network_id, tld, maybe_filter)?;
+
+        let mut results = Vec::new();
+        for deployed_protocol in protocols {
+            let empty_label_hash = deployed_protocol
+                .protocol
+                .info
+                .protocol_specific
+                .empty_label_hash();
+
+            let domain_name = DomainName::new(name_with_tld, empty_label_hash)?;
+            results.push(DomainNameOnProtocol::from_str(
+                &domain_name.name,
+                deployed_protocol,
+            ));
+        }
+
+        let results = results
+            .into_iter()
+            .filter_map(|result| result.ok())
+            .collect::<Vec<_>>();
+        if results.is_empty() {
+            return Err(ProtocolError::InvalidName {
+                name: name_with_tld.to_string(),
+                reason: "No matching protocols for given TLD".to_string(),
+            });
+        }
+        Ok(results)
+    }
+
+    pub fn lookup_names_options_in_network(
+        &self,
+        name: &str,
+        network_id: i64,
+        maybe_filter: Option<NonEmpty<String>>,
+    ) -> Result<Vec<DomainNameOnProtocol>, ProtocolError> {
+        if name.contains('.') {
+            let direct = self.fetch_domain_options(name, network_id, maybe_filter)?;
+
+            if direct.is_empty() {
+                Err(ProtocolError::InvalidName {
+                    name: name.to_string(),
+                    reason: "No matching protocols for given TLD".to_string(),
+                })
```

Comment:
you check that your results `.is_empty()` two times in `lookup_names_options_in_network` and in `fetch_domain_options`, why?

consider using [NonEmpty](https://docs.rs/nonempty/latest/nonempty/) vector here if you want to ensure that your vector has something inside. I already used in codebase